Feature/iodevice body

Garcia6l20 opened the pull request:
This PR enables `QAbstractHttpServer` to handle `QIODevice` instead of `QByteArray` to manage its body.
This allow us to provide direct forwarding body data to any `QIODevice` (eg.: `QFile`)

Commit:
CMake setup

- fix: header not included when compiling through CMake

Commit:
CMake setup

- fix: header not included when compiling through CMake

Commit:
`final` modifier removed

- allow `QHttpServer` to be subclassed

msvetkin committed:
QHttpServerResponse: Extend the API

Add new API for HTTP headers manipulations.
Add QHttpServerResponse::write function which will
allow to write custom response objects in a future.

Fixes: QTBUG-76933
Change-Id: I744303be1b517c07f698c4a3dd2c4296f77e3b03
Reviewed-by: Tasuku Suzuki <tasuku.suzuki@tqcs.io>
Reviewed-by: Edward Welbourne <edward.welbourne@qt.io>
Reviewed-by: Mårten Nordheim <marten.nordheim@qt.io>

Commit:
QIODevice based body management

Commit:
Fix: seek body postion upon body read

Commit:
CMake: adding unit tests

Commit:
Adding request param to QHttpServerRouterRule::createBodyDevice

Commit:
fix: missing AUTO_MOC

Commit:
Option QTHTTPSERVER_TESTS_EN

Commit:
Adding direct file upload test

- plain text error on http_parser error.

BelgacemNajib commented on src/httpserver/qhttpserver_p.h:
```diff
@@ -56,7 +56,7 @@ class QHttpServerPrivate: public QAbstractHttpServerPrivate
     Q_DECLARE_PUBLIC(QHttpServer)
 
 public:
-    QHttpServerPrivate() = default;
```
why removing this ?

Garcia6l20 replied:
We need the server pointer to be able to call `server->createBodyDevice()`

BelgacemNajib commented on src/httpserver/qabstracthttpserver.cpp:
```diff
@@ -127,8 +129,12 @@ void QAbstractHttpServerPrivate::handleReadyRead(QTcpSocket *socket,
         Q_EMIT q->missingHandler(*request, socket);
 }
 
+QIODevice* QAbstractHttpServer::createBodyDevice(const QHttpServerRequest& /*request*/) {
```
shared pointer if possible

Garcia6l20 replied:
I would prefer use the Qt's onwership memory management @lmussier An advice ?

lmussier replied:
Agree with sylvain, we are in Qt world.

Garcia6l20 commented on src/httpserver/qhttpserver.cpp:
```diff
@@ -61,7 +65,7 @@ Q_LOGGING_CATEGORY(lcHS, "qt.httpserver");
 
 QHttpServer::QHttpServer(QObject *parent)
-    : QAbstractHttpServer(*new QHttpServerPrivate, parent)
+    : QAbstractHttpServer(*new QHttpServerPrivate(this), parent)
```
What is that ??? Memory leak ???

Garcia6l20 replied:
Valgrind does not complain... I'll check what will happend with that in Qt's sources

Garcia6l20 replied:
Ok... No memleak here.
QObject takes ownership of the private data, by taking the pointer of the given reference

lmussier commented on tests/auto/qhttpserver/tst_qhttpserver.cpp:
```diff
@@ -101,7 +119,7 @@ private slots:
     int data = 10;
 
-    CustomArg() {} ;
+    CustomArg() {}
```
= default ?

Garcia6l20 replied:
It was just a code cleanup... Effect is exatly the same...

lmussier commented on src/httpserver/qabstracthttpserver.cpp:
```diff
@@ -127,8 +129,12 @@ void QAbstractHttpServerPrivate::handleReadyRead(QTcpSocket *socket,
+
 QAbstractHttpServer::QAbstractHttpServer(QObject *parent)
-    : QAbstractHttpServer(*new QAbstractHttpServerPrivate, parent)
+    : QAbstractHttpServer(*new QAbstractHttpServerPrivate(this), parent)
```
This is some kind of mutual dependency. Generally the symptom of a design problem.
The destructor might be tricky :s

Garcia6l20 replied:
Yes, this is quite ugly... 
Most of the process is done within the private stufs an I need a reference to the actual non-private stuf to be able to call `createBodyDevice()` (from`QHttpServerRequesPrivate` where the actual parsing is done)

Garcia6l20 replied:
I would prefer passing the IO device directly in `QHttpServerRequest` ctor, but we might need some informations from the acutal request eg.: filename. So I just cannot.

Garcia6l20 replied:
The goal is to be able to call:
```cpp
server->createBodyDevice(request);
```
from `QHttpServerRequestPrivate` (where http parsing is done)

The only way I can see that we'll be able to remove mutual dependency is to use `std::function`:
Ctor example:
```cpp
QHttpServerRequest(std::function<QIODevice*(QHttpServerRequest&)> ioDeviceProvider)
```

lmussier commented on src/httpserver/qabstracthttpserver.h:
```diff
@@ -75,8 +75,11 @@ class Q_HTTPSERVER_EXPORT QAbstractHttpServer : public QObject
     static QHttpServerResponder makeResponder(const QHttpServerRequest &request,
                                               QTcpSocket *socket);
 
+    virtual QIODevice* createBodyDevice(const QHttpServerRequest& request);
```
For me this is not the right place, the server should not have to be aware of how requests are handled.
All should be done around QHttpServerRequest. The only thing I see on httpserver is the possibility to configure where to put uploaded files

Garcia6l20 replied:
But dont we want to bind specific RouterRule to a spécific IODevice ?

lmussier commented on src/httpserver/qabstracthttpserver.h:
```diff
@@ -75,8 +75,11 @@ class Q_HTTPSERVER_EXPORT QAbstractHttpServer : public QObject
+
 private:
     Q_DECLARE_PRIVATE(QAbstractHttpServer)
+    friend class QHttpServerRequestPrivate;
```
This is the goes with the above remark, we should not have to do that, responsibilities have to be clearly separated between server and request

lmussier commented on src/httpserver/qhttpserver.cpp:
```diff
@@ -141,4 +145,9 @@ bool QHttpServer::handleRequest(const QHttpServerRequest &request, QTcpSocket *s
     return d->router.handleRequest(request, socket);
 }
 
+QIODevice* QHttpServer::createBodyDevice(const QHttpServerRequest& request) {
```
For me this is not the role of the server

Garcia6l20 committed:
HTTPS support

Added new `QAbstractHttpServer::sslSetup` which enables HTTPS usage.
Added new `QSslServer` which inherits from `QTcpServer` and configures
incoming TCP clients to use SSL.

[ChangeLog][QHttpServer][Https support] Https support added to
QAbstractHttpServer class

Change-Id: I536cf48b86b246e3f4b9d960f793b93670afe06f
Reviewed-by: Edward Welbourne <edward.welbourne@qt.io>
Reviewed-by: Mårten Nordheim <marten.nordheim@qt.io>
Reviewed-by: Guy Poizat <gerrit.qt@gmail.com>
Reviewed-by: Mikhail Svetkin <mikhail.svetkin@gmail.com>

Garcia6l20 committed:
Fix memory leak in QHttpServerRequest

Using QSharedPointer rather than raw pointer for private data

[ChangeLog][QHttpServerRequest][Fix QHttpServerRequest memory leak] Fixed: memory leak in QHttpServerRequest

Change-Id: I419ec22e56f199677d058938cee046a60a3c2450

Garcia6l20 committed:
removed useless dependencies

Change-Id: Ib26a4e318897367516cfbd3c0b59f5c46b584242

Garcia6l20 committed:
ssl errors handling

Change-Id: Ic987dff26c6cc30c3146d59f0ef473e2e6f115d4

Garcia6l20 committed:
fix request cleanup

Change-Id: I5159faeff38c00bba71d884505ef1bbd5dcb44b0

Garcia6l20 committed:
test fixed

Change-Id: Ia247bc076ad036acdd331098786330c6d27dd493
